### ISSUE TITLE
Chat - Pressing CMD+K when Global create is open closes the menu and doesn’t open search

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#24256]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
and log in
2. From LHN press the plus icon
3. From the keyboard press CMD+K (or CTRL+K on Windows)

## Expected Result:
The Global create menu closes and the Right hand modal opens with search

## Actual Result:
The Global create menu closes but the Right hand modal doesn't open with search

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.51.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/9fffbdbe-2dc0-401f-b078-8eb616338bf9

https://github.com/Expensify/App/assets/78819774/068fa70e-ad0c-4c69-80e0-bba271163fd8

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause -Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Beamanator** [comment]
I wouldn't consider this a blocker - cc @puneetlath 

And def can be worked externally, marking as such

**Comment by alanleungcn** [comment]
Contributor details
Your Expensify account email: alan@alanleungcn.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f5c90a04dc1b08b8

**Comment by Thanos30** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
CMD+K keyboard shortcut doesn't work when `FloatingActionButtonAndPopover` is open. Also happening with CMD+SHIFT+K

### What is the root cause of that problem?
The root cause of this problem is in the way we subscribe the keyboard shortcuts in `AuthScreens.js`:
```
this.unsubscribeSearchShortcut = KeyboardShortcut.subscribe(
            searchShortcutConfig.shortcutKey,
            () => {
                Modal.close(() => {
                    if (Navigation.isActiveRoute(ROUTES.SEARCH)) {
                        return;
                    }
                    return Navigation.navigate(ROUTES.SEARCH);
                });
            },
            searchShortcutConfig.descriptionKey,
            searchShortcutConfig.modifiers,
            true,
        );
```.
The shortcut subscription for creating a new group is similar. As you can see, we are sending the callback inside the close function of the Modal:

```
function close(onModalCloseCallback, isNavigating = true) {
    if (!closeModal) {
        // If modal is already closed, no need to wait for modal close. So immediately call callback.
        if (onModalCloseCallback) {
            onModalCloseCallback();
        }
        onModalClose = null;
        return;
    }
    onModalClose = onModalCloseCallback;
    closeModal(isNavigating);
}
```

At this point, our shortcut is closing the Modal, which triggers the Modal's closing function, which is set on `FloatingActionButtonAndPopover`, and it's overriding our shortcut.

### What changes do you think we should make in order to solve the problem?
Since we want our keyboard shortcuts to close the Modals and navigate to the appropriate location (Search or New group), it's better to respect the `close` function of each modal, and then proceed with the navigation.

So, in `AuthScreens.js` , we will make the subscriptions like this:

```
this.unsubscribeSearchShortcut = KeyboardShortcut.subscribe(
            searchShortcutConfig.shortcutKey,
            () => {
                Modal.close();
                if (Navigation.isActiveRoute(ROUTES.SEARCH)) {
                    return;
                }
                Navigation.navigate(ROUTES.SEARCH);
            },
            searchShortcutConfig.descriptionKey,
            searchShortcutConfig.modifiers,
            true,
        );
```
, where we are first running the close() function, and later navigate, so we don't overlap the onClose functions.

**Comment by SosenWiosen** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
CMD+K doesn't navigate to search when FloatingActionButtonAndPopover is open.

### What is the root cause of that problem?
FloatingActionButtonAndPopover uses PopoverWithoutOverlay. In PopoverWithoutOverlay/index.js   Modal.onModalDidClose() isn't called when the Popover gets closed. 

`React.useEffect(() => {
        if (props.isVisible) {
            props.onModalShow();
            onOpen({
                ref,
                close: props.onClose,
                anchorRef: props.anchorRef,
            });
        } else {
            props.onModalHide();
            close(props.anchorRef);
        }
        Modal.willAlertModalBecomeVisible(props.isVisible);
        Modal.setCloseModal(props.isVisible ? () => props.onClose(props.anchorRef) : null);

        // We want this effect to run strictly ONLY when isVisible prop changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.isVisible]);

    if (!props.isVisible) {
        return null;
    }`
    
    This is also different to BaseModal.js, where this.hideModal is called in componentWillUnmount.
    `componentWillUnmount() {
        // Only trigger onClose and setModalVisibility if the modal is unmounting while visible.
        if (this.props.isVisible) {
            this.hideModal(true);
            Modal.willAlertModalBecomeVisible(false);
        }

        // To prevent closing any modal already unmounted when this modal still remains as visible state
        Modal.setCloseModal(null);
    }`
 
### What changes do you think we should make in order to solve the problem?
We should change the useEffect in PopoverWithoutOverlay.js to call    Modal.onModalDidClose(); (the same logic as in BaseModal.js).

`React.useEffect(() => {
        if (props.isVisible) {
            props.onModalShow();
            onOpen({
                ref,
                close: props.onClose,
                anchorRef: props.anchorRef,
            });
        } else {
            props.onModalHide();
            close(props.anchorRef);
            Modal.onModalDidClose();
        }
        Modal.willAlertModalBecomeVisible(props.isVisible);
        Modal.setCloseModal(props.isVisible ? () => props.onClose(props.anchorRef) : null);

        // We want this effect to run strictly ONLY when isVisible prop changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.isVisible]);`


<details>
<summary>Video</summary>

https://github.com/Expensify/App/assets/43684335/c8580404-da2a-4df3-af9c-1d6f22996b8f

</details>
### What alternative solutions did you explore? (Optional)

**Comment by SosenWiosen** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
CMD+K doesn't navigate to search when FloatingActionButtonAndPopover is open.

### What is the root cause of that problem?
FloatingActionButtonAndPopover uses PopoverWithoutOverlay. In PopoverWithoutOverlay/index.js   Modal.onModalDidClose() isn't called when the Popover gets closed. 

`React.useEffect(() => {
        if (props.isVisible) {
            props.onModalShow();
            onOpen({
                ref,
                close: props.onClose,
                anchorRef: props.anchorRef,
            });
        } else {
            props.onModalHide();
            close(props.anchorRef);
        }
        Modal.willAlertModalBecomeVisible(props.isVisible);
        Modal.setCloseModal(props.isVisible ? () => props.onClose(props.anchorRef) : null);

        // We want this effect to run strictly ONLY when isVisible prop changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.isVisible]);

    if (!props.isVisible) {
        return null;
    }`
    
    This is also different to BaseModal.js, where this.hideModal is called in componentWillUnmount.
    `componentWillUnmount() {
        // Only trigger onClose and setModalVisibility if the modal is unmounting while visible.
        if (this.props.isVisible) {
            this.hideModal(true);
            Modal.willAlertModalBecomeVisible(false);
        }

        // To prevent closing any modal already unmounted when this modal still remains as visible state
        Modal.setCloseModal(null);
    }`
 
### What changes do you think we should make in order to solve the problem?
We should change the useEffect in PopoverWithoutOverlay.js to call    Modal.onModalDidClose(); (the same logic as in BaseModal.js).

`React.useEffect(() => {
        if (props.isVisible) {
            props.onModalShow();
            onOpen({
                ref,
                close: props.onClose,
                anchorRef: props.anchorRef,
            });
        } else {
            props.onModalHide();
            close(props.anchorRef);
            Modal.onModalDidClose();
        }
        Modal.willAlertModalBecomeVisible(props.isVisible);
        Modal.setCloseModal(props.isVisible ? () => props.onClose(props.anchorRef) : null);

        // We want this effect to run strictly ONLY when isVisible prop changes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.isVisible]);`


<details>
<summary>Video</summary>

https://github.com/Expensify/App/assets/43684335/c8580404-da2a-4df3-af9c-1d6f22996b8f

</details>
### What alternative solutions did you explore? (Optional)

**Comment by dangrous** [comment]
I think we can consider it a regression so probably not eligible to be a separate issue with separate payment... But I think we also agreed that we wouldn't cut the payment of the original PR for regressions since it was such a big piece of work. Does that make sense?

**Comment by Thanos30** [comment]
Pardon my intrusion, but this issue seems to be more on the handling of the keyboard shortcut and less on the Modal re-designs. As I stated on my proposal, the keyboard shortcut should respect the Modal's onClose function and not try to overlap it.

Wouldn't that be considered a different issue?

Thank you in advance! 🙏

**Comment by SosenWiosen** [comment]
@Thanos30  I don't agree with you that it's caused by values overlapping.  See my proposal.

**Comment by SosenWiosen** [comment]
I agree that the code in keyboard shortcuts should be more consistent, but this isn't the root cause of the problem here. 
I think that in your last comment you are confusing  Modal.onModalDidClose() with the callbacks from FloatingActionButtonAndPopover. While the latter get's called, the former doesn't.
PopoverWithoutOverlay should be consistent with how the modals behave, and should call Modal.onModalDidClose();

I agree with @aimane-chnaif  that this is a regression.

**Comment by Thanos30** [comment]
Got it, thank you for the clarification @SosenWiosen, you might be right about that.

**Comment by zanyrenney** [comment]
Cool, so if this is a regression @allroundexperts you are working on a PR for this?

**Comment by zanyrenney** [comment]
thank you for the update!

**Comment by mananjadhav** [comment]
Considering this is a regression and @rushatgabhane @allroundexperts are handling it, I am unassigning myself.

**Comment by Beamanator** [comment]
Looks like this was potentially linked as a regression (see link in [this comment](https://github.com/Expensify/App/issues/24256#issuecomment-1698170677)) but that was taken back as the PR for this issue actually fixes a few regressions from a different PR.

So I think we're good to pay this out as normal, though @dangrous it seems you had some thoughts about the original PR that caused these regressions? I think we should be good to pay this out as normal, right?

**Comment by dangrous** [comment]
I thiiiink we were deciding that no payment needed for this one since it was a regression on that other PR - but lemme know if you disagree!
